content/blog/eng/functional-testing-symfony-guzzle.md

```diff
@@ -0,0 +1,319 @@
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Handler\MockHandler;
+use GuzzleHttp\HandlerStack;
+use GuzzleHttp\Psr7\Response;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Exception\RequestException;
+
+// Create a mock and queue two responses.
+$mock = new MockHandler([
+    new Response(200, ['X-Foo' => 'Bar']),
+    new Response(202, ['Content-Length' => 0]),
+    new RequestException("Error Communicating with Server", new Request('GET', 'test'))
+]);
+
+$handler = HandlerStack::create($mock);
+$client = new Client(['handler' => $handler]);
+
+// The first request is intercepted with the first response.
+echo $client->request('GET', '/')->getStatusCode();
+//> 200
+// The second request is intercepted with the second response.
+echo $client->request('GET', '/')->getStatusCode();
+//> 202
+```
+
+# A real use case
+But how can we test our actual symfony controller 
+that must handle all the response? Let's suppose we have a controller like this: 
+it returns a JsonResponse with a different message and status code based on what 
+it gets from the APIs
+
+```php
+<?php
+
+use Symfony\Component\HttpFoundation\JsonResponse;
+use MyNamespace\MyApiClient;
+
+public function apiControllerAction(MyApiClient $client)
+    {
+     
+      $response = $client->getFoo();
+
+      if($response->getStatusCode()===200){
+            return new JsonResponse(
+                [
+                    'status'=>'ok',
+                    'message' => $response->getBody()
+                ],
+                 200
+            );
+        }
+
+        if($response->getStatusCode()===404){
+            return new JsonResponse(
+                [
+                   'status' => 'error',
+                   'message' => 'foo not found'
+                ],
+                 404
+            );
+        }
+
+        return new JsonResponse(
+            [
+                'status'=>'error',
+                'message' => 'something went wrong'
+            ],
+            500
+        );
+
+    }
+```
+
+MyApiClient is done this way: it has a constructor that initialize the http client 
+and a method who retrieves foo from the API.
+
+```php
+<?php
+
+namespace MyNamespace;
+use Psr\Http\Client\ClientInterface;
+
+class MyApiClient
+{
+    public function __construct(ClientInterface $client)
+    {
+        $this->client = $client;
+    }
+
+    public function getFoo()
+    {
+        return $this->get('https://api-url/foo');
+    }
+
+}
+```
+
+Symfony will automatically inject MyApiClient in our controller, but to make sure 
+it will inject ClientInterface in MyApiClient we have to modify services.yaml to 
+look like this:
+
+```yaml
+Psr\Http\Client\ClientInterface: '@psr18.client'
+
+  psr18.client:
+    class: GuzzleHttp\Client
+```
+
+This way we are telling to symfony inject Guzzle Http Client in MyApiClient
+
+# Road to testing
+Now we have to create our controller's functional test. It will expect 3 different 
+responses according to API responses.
+But first…
```

Comment:
It is not so clear. Consider write a full sentence explaining plainly your point.
Here's an example (text in italic is just placeholder for the effective words):
> Before _creating our test_, however, a further/preliminary _step_ is needed.